crates/dbsp/src/operator/sample.rs

```diff
@@ -146,10 +146,17 @@ where
                 if sample_size <= num_quantiles {
                     sample
                 } else {
+                    let mut keys = Vec::with_capacity(sample_size);
```

Comment:
It does work `O(s^2)` where `s` is the ultimate sample size. I don't think it increases the overall big-O of the sampling operation because the code that generates the sample that it re-samples also does `O(s^2)` work.

Reply:
Well, I replaced all the batches with the new persistent batches, but I see what you mean that in this case it is unnecessary. I can change it back for this one, and for any others that we notice.

Reply:
I was pretty much concentrating on "let's swap out the implementation and make sure it's a drop-in replacement", so I guess I was blind to cases where replacement wasn't needed.

Reply:
I added a commit that uses the non-persistent zset for this.

Reply:
Making sure that persistent batches are a drop-in replacement for the in-memory ones is a good idea, but I don't think that we want to run the system this way.  I think that only batches in traces need to be persistent.  Batches computed by operators on the fly should stay in-memory.  This is something we didn't think through sufficiently.  Let's have an offline discussion.

I didn't realize that this PR switches to persistent batches throughout. If it does, flip that switch back before merging.

Reply:
That's what I thought originally too, but I don't think it's true.
This is necessary for any collection that does not fit in RAM.

Reply:
And during backfill any collection can be very large.

Reply:
> Making sure that persistent batches are a drop-in replacement for the in-memory ones is a good idea, but I don't think that we want to run the system this way. I think that only batches in traces need to be persistent. Batches computed by operators on the fly should stay in-memory. This is something we didn't think through sufficiently. Let's have an offline discussion.

I do need that discussion.

> I didn't realize that this PR switches to persistent batches throughout. If it does, flip that switch back before merging.

By design, it's easy to flip back, so I'll do that.

Reply:
> And during backfill any collection can be very large.

True, but doing this for all batches can have massive impact on performance.

Reply:
you want some mechanism like swapping, where if a collection exceeds a certain size you start using storage

Reply:
It should be possible to build an implementation that starts out in-memory and then spills to disk at a certain size. That can be a next step.